src/agg_bookend_optimize.c

```diff
@@ -0,0 +1,678 @@
+ *		Recursively scan the Aggref nodes in an expression tree, and check
+ *		that each one is a FIRST/LAST aggregate.  If so, build a list of the
+ *		distinct aggregate calls in the tree.
+ *
+ * Returns TRUE if a non-FIRST/LAST aggregate is found, FALSE otherwise.
+ * (This seemingly-backward definition is used because expression_tree_walker
+ * aborts the scan on TRUE return, which is what we want.)
+ *
+ * Found aggregates are added to the list at *context; it's up to the caller
+ * to initialize the list to NIL.
+ *
+ * This does not descend into subqueries, and so should be used only after
+ * reduction of sublinks to subplans.  There mustn't be outer-aggregate
+ * references either.
+ */
+static bool
+find_first_last_aggs_walker(Node *node, List **context)
+{
+	if (node == NULL)
+		return false;
+	if (IsA(node, Aggref))
+	{
+		Aggref	   *aggref = (Aggref *) node;
+		Oid			aggsortop;
+		TargetEntry *value;
+		TargetEntry *sort;
+		MinMaxAggInfo *mminfo;
+		ListCell   *l;
+		FirstLastAggInfo *fl_info;
+		Oid			sort_oid;
+
+		Assert(aggref->agglevelsup == 0);
+		if (list_length(aggref->args) != 2)
+			return true;		/* it couldn't be first/last */
+
+		/*
+		 * ORDER BY is usually irrelevant for FIRST/LAST, but it can change
+		 * the outcome if the aggsortop's operator class recognizes
+		 * non-identical values as equal.  For example, 4.0 and 4.00 are equal
+		 * according to numeric_ops, yet distinguishable.  If FIRST() receives
+		 * more than one value equal to 4.0 and no value less than 4.0, it is
+		 * unspecified which of those equal values FIRST() returns.  An ORDER
+		 * BY expression that differs for each of those equal values of the
+		 * argument expression makes the result predictable once again.  This
+		 * is a niche requirement, and we do not implement it with subquery
+		 * paths. In any case, this test lets us reject ordered-set aggregates
+		 * quickly.
+		 */
+		if (aggref->aggorder != NIL)
+			return true;
+		/* note: we do not care if DISTINCT is mentioned ... */
+
+		/*
+		 * We might implement the optimization when a FILTER clause is present
+		 * by adding the filter to the quals of the generated subquery.  For
+		 * now, just punt.
+		 */
+		if (aggref->aggfilter != NULL)
+			return true;
+
+		/* We sort by second argument (eg. time) */
+		sort_oid = lsecond_oid(aggref->aggargtypes);
+
+		/* Let's set the sort operator oid */
+		if (is_first_agg(aggref))
+			aggsortop = OpernameGetOprid(list_make1(makeString("<")), sort_oid, sort_oid);
+		else if (is_last_agg(aggref))
+			aggsortop = OpernameGetOprid(list_make1(makeString(">")), sort_oid, sort_oid);
+		else
+		{
+			elog(WARNING, "Failed to resolve sort operator for sort oid: %d", sort_oid);
+			return true;		/* not a first/last aggregate */
+		}
+
+		/* Used in projection */
+		value = (TargetEntry *) linitial(aggref->args);
+		/* Used in ORDER BY */
+		sort = (TargetEntry *) lsecond(aggref->args);
+
+		if (contain_mutable_functions((Node *) value->expr))
```

Comment:
do we care if value is mutable. Def we care about sort but not sure about value.